Java API: Fix OrtSession.close() crashing in multi-threaded environments

yuzawa-san opened the pull request:
**Description**: This change makes all of the classes backed by native objects extends a new NativeObject class which properly handles the the safe access and execution of native methods. Specifically, any native objects needed for a specific native method must be not closed. Additionally, an object's close method to wait until the last usage of any active reference is completed. Failure to do those things may result in a native method being invoked with closed native handle. Or a native close occurring during an active native execution. I was inspired by tensorflow's JNI https://github.com/tensorflow/tensorflow/blob/master/tensorflow/java/src/main/java/org/tensorflow/Graph.java#L359 Like their implementation, I utilized a lot of try-with-resources. I did not like the amount of locking in their implementation, so I decided to use reference counts using AtomicInteger to ensure consistency using CAS rather than synchronization. I also did some general cleanup:
* Immutable maps and collections.
* Reading session information only at construction time.
* Pre-sizing maps and collections.
* Update gitignore for IDE files.
* Update javadoc config to only show public methods.
* Cleaned up default allocator and session options in OrtEnvironment. This did require moving some OrtAllocator-specific code in the C code to a different file.

***Thread-Safety Description***

Here is some pseudocode of what is going on in NativeObject:

```
class NativeObject {
  constructor() {
  increment referenceCount
  }

  close() {
   if closed do not close again
   decrement referenceCount
   if the new value is greater than zero then some other thread is still using a reference (at least one NativeReference has not been closed) so wait for that last alive one to notify this thread before proceeding
   else the new value is less than or equal to zero, so there are no active references (all instances of NativeReference have been closed) so we do not need to wait
   free up native resources
  }
  
  class NativeReference {
    constructor() {
      increment referenceCount
      if the old value was less than or equal to zero indicating the NativeObject was closed then release() and throw an exception. release() needs to be called to restore the old value by decrementing the referenceCount. also the exception means that close() won't be called.
      else proceed as normal and the release() will be called in the close()
    }
    
    release() {
      decrement referenceCount
      if the new value is zero then this was the last active reference so notify the wait in NativeObject.close()
      else do nothing since NativeObject is still alive
    }
    
    close() {
      release() if NativeReference has not been closed
    }
  }
}
```


**Motivation and Context**
- I intend on using onnxruntime in a multithreaded environment. The session is created in one thread, then a bunch of other threads see the reference to the OrtSession and do model evaluations in those other threads. Periodically, a new model is loaded and a new OrtSession is created. The reference to the active session is swapped out for the new OrtSession, and then that OrtSession would be closed. However the current codebase does not handle this situation gracefully or safely. Inferences could still be occurring in the run() method, when the OrtSession.close() method is invoked. Depending on the interleaving across threads this could produce a segmentation fault. At the minimum, any objects used after closure should throw an exception. And more importantly, the close() method should be able to block until the last usage is complete, while simultaneously locking out any new attempts to use the object.
- I also updated the OrtEnvironment semantics. While in most cases, a single environment is created and used for the duration of the application. I did add the ability to close the OrtEnvironment. In that case, a subsequent call to getEnvironment() would not return the existing environment, but rather a new one.

Craigacp commented on java/src/main/java/ai/onnxruntime/OnnxMap.java:
```diff
@@ -104,12 +94,12 @@ public static OnnxMapValueType mapFromOnnxJavaType(OnnxJavaType type) {
    *
    * @param nativeHandle The reference to the native map object.
-   * @param allocatorHandle The reference to the allocator that created the map.
+   * @param allocator The allocator used to when extracting data from this object.
```
"The allocator used to when" -> "The allocator used when"

Craigacp commented on java/src/main/java/ai/onnxruntime/OrtAllocator.java:
```diff
@@ -4,45 +4,32 @@
-      OnnxRuntime.init();
-    } catch (IOException e) {
-      throw new RuntimeException("Failed to load onnx-runtime library", e);
+      DEFAULT_ALLOCATOR = new OrtAllocator(getDefaultAllocator(OnnxRuntime.ortApiHandle), true);
```
It wasn't clear to me if the default allocator could exist separately from the OrtEnvironment. It looks like this allocator has a different lifespan, have you found any issues from that?

yuzawa-san replied:
I checked the C API example to be sure: https://github.com/microsoft/onnxruntime/blob/master/csharp/test/Microsoft.ML.OnnxRuntime.EndToEndTests.Capi/C_Api_Sample.cpp#L64

The allocator seems separate from the environment, so it appears to be safe to move this.

Craigacp replied:
Ok. We should get someone from MS to check this. I think you're right, it might have been the initial version of this that I wrote against the 0.5.0 API which had allocator issues.

yuslepukhin replied:
Default allocator is a singleton. There is a PR in works that is making things more evident in C#.

---
In reply to: [445867761](https://github.com/microsoft/onnxruntime/pull/4329#discussion_r445867761) [](ancestors = 445867761)

Craigacp commented on java/src/main/java/ai/onnxruntime/OrtSession.java:
```diff
@@ -441,7 +484,10 @@ private native OnnxModelMetadata constructMetadata(
-  public static class SessionOptions implements AutoCloseable {
+  public static class SessionOptions extends NativeObject {
+
+    static final OrtSession.SessionOptions DEFAULT_SESSION_OPTIONS =
```
Is this a good idea? I think the session options isn't designed to be shared between different sessions. I realise the default one doesn't seem to currently hold any shared resources, but that's not to say it won't in the future, and such a change wouldn't be visible from the Java API.

yuzawa-san replied:
Yeah I'm fine creating a new session options each time, since they are very minimal.

Craigacp commented on java/src/main/java/ai/onnxruntime/OrtSession.java:
```diff
@@ -846,13 +911,11 @@ private native long registerCustomOpLibrary(long apiHandle, long nativeHandle, S
-     * The order they care called indicates the preference order as well. In other words call this method
-     * on your most preferred execution provider first followed by the less preferred ones.
-     * If none are called Ort will use its internal CPU execution provider.
+     * functions to enable them in the session: OrtSessionOptionsAppendExecutionProvider_CPU
```
The reformatting broke this list here, please revert it.

Craigacp commented on java/src/main/native/ai_onnxruntime_OrtSession.c:
```diff
@@ -235,16 +235,18 @@ JNIEXPORT jobjectArray JNICALL Java_ai_onnxruntime_OrtSession_getOutputInfo
     OrtAllocator* allocator = (OrtAllocator*) allocatorHandle;
     OrtSession* session = (OrtSession*) sessionHandle;
     OrtRunOptions* runOptions = (OrtRunOptions*) runOptionsHandle;
+    jsize numInputs = (*jniEnv)->GetArrayLength(jniEnv, inputNamesArr);
```
Isn't it faster to pass in the longs rather than make an upcall into the JVM?

Craigacp commented on java/src/test/java/ai/onnxruntime/InferenceTest.java:
```diff
@@ -289,32 +349,18 @@ public void partialInputsTest() throws OrtException {
-      inputMap.clear();
-      requestedOutputs.clear();
-      */
+      /*
```
The reformat broke this comment, please revert it.

yuzawa-san commented:
@Craigacp I believe I addressed your comments. The formatting changes were done automatically by the spotless plugin. I did change how they were formatted to convey the original layout better.

I do admit this is a little more complex, but I think it shields developers (Java devs especially) from the gnarly world of segmentation faults, JNI race conditions, and other nasty crashes. I'd much prefer to have some exceptions or blocking closes rather than whole process crashes. I would consider tensorflow to be similar to this project and they have it.

I'm going to think about the RunOptions idea. I think it could work in conjunction with what I currently have. What do you think of adding a terminate() (or some other named) method to OrtSession which would do what you mentioned. So if somebody was in a multithreaded environment they would call session.terminate() and then session.close() and theoretically the close() would not block.

Craigacp commented:
Hmm, it's weird that spotless wasn't changing the format when I ran it. Are you using a different one to the one pulled in by gradle?

yuzawa-san commented:
I have been running gradle's `spotlessApply` task to reformat.

I added a OrtSession.terminate() method which can be used optionally in cases when a developer is in a multi-threaded environment and they want the subsequent call to OrtSession.close() to not block. I tried to document this. I used your idea of using a default RunOptions.

It looks like TF does block on close() in the same manner as this implementation. In both cases the blocking is only done in cases when there are extant usages at the time of closing, specifically in multi-threaded cases. Single-threaded developers need not worry about blocking.

Unrelated, but I also reduced the number of JNI calls done on session creation which seemed redundant. If there was a specific reason the quantity, names, and NodeInfos are 3 different JNI calls, let me know and I can leave it alone.

Craigacp commented on java/src/main/java/ai/onnxruntime/OrtSession.java:
```diff
@@ -199,7 +194,7 @@ public long getNumOutputs() {
    */
   public Result run(Map<String, OnnxTensor> inputs) throws OrtException {
-    return run(inputs, outputNamesArray, null);
+    return run(inputs, outputNamesArray, defaultRunOptions);
```
I don't know if the RunOptions instance can be shared between concurrent runs. @pranavsharma Could someone from MS tell us about the thread safety and shareability of these classes?

yuzawa-san replied:
I don't know a lot about CPP thread safety, but i think these are some reads and writes of that terminate flag: 
https://github.com/microsoft/onnxruntime/blob/0638565fe0d6149d1512ec3e63512e6ea13743a3/include/onnxruntime/core/framework/run_options.h#L23
https://github.com/microsoft/onnxruntime/blob/7c89f38a346bcfc4494884d8278ba838495e9575/onnxruntime/core/framework/run_options.cc#L48
https://github.com/microsoft/onnxruntime/blob/274e6b41535f79808e2c6d52406851479ab6dca0/onnxruntime/core/framework/parallel_executor.cc#L129

In this case the terminate flag would be set in another thread from the model execution.

Craigacp replied:
Yeah, I'm fairly confident the terminate functionality would be fine, but my worry is if the C++ devs consider this to be a class that is only bound to a single call to run, then they might refactor things in a way that break us without a C API update. Which is why I'd like someone from the core team to check (or tell us where we could find that out from the docs).

Craigacp replied:
Given it's a small object, with presumably low overhead (apart from the JNI call to create it) it might be worth manufacturing one each time, as that would avoid this penalty at the cost of slightly increasing latency (for the JNI) and a marginal increase in garbage (for the Java object).

Craigacp replied:
Ah, I see you did that in the latest change.

yuzawa-san replied:
What do the maintainers think? Is it possible to setTerminate in one thread and safely stop execution in another thread?

yuslepukhin replied:
All Run() interfaces in Session object take RunOptions by const ptr/reference. The input parameter is not modified.
The references above has to do either with the interfaces that allow you change the default values of RunOptions before running inference.
The last reference has to do with the parallel executor member and not with RunOPtions. It it safe to say that it has nothing to do with Java wrappers.

---
In reply to: [453974752](https://github.com/microsoft/onnxruntime/pull/4329#discussion_r453974752) [](ancestors = 453974752)

Craigacp commented on java/src/main/java/ai/onnxruntime/OrtSession.java:
```diff
@@ -386,19 +383,43 @@ public String endProfiling() throws OrtException {
+   * #close()} to prevent the normal behavior of blocking until pending model evaluations are
+   * completed.
+   *
+   * <p>Note: This will only terminate calls to {@link #run(Map)} and {@link #run(Map,Set)}. It will
```
I had envisioned that the OrtSession would track all the currently executing runs and then call terminate on all the RunOptions that are there. Poke them into a HashSet on the way through and remove them on the way out.

yuzawa-san replied:
Yeah that makes more sense I'll update it

Craigacp commented on java/src/main/java/ai/onnxruntime/OrtSession.java:
```diff
@@ -109,7 +110,7 @@ private OrtSession(long sessionHandle, OrtEnvironment environment, OrtAllocator
     this.environment = environment;
     this.allocator = allocator;
-    this.defaultRunOptions = new RunOptions();
+    this.activeRunOptions = ConcurrentHashMap.newKeySet();
```
Doesn't this set need a default mapped value to work (see https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.KeySetView.html#add-K-).

yuzawa-san replied:
It should work as it. I tested with:
```java
Set<String> set = ConcurrentHashMap.newKeySet();
System.out.println(set);
set.add("foo");
System.out.println(set);
set.add("bar");
System.out.println(set);
set.add("foo");
System.out.println(set);
set.add("baz");
System.out.println(set);
set.remove("foo");
System.out.println(set);
```
and got:
```
[]
[foo]
[bar, foo]
[bar, foo]
[bar, foo, baz]
[bar, baz]
```

Craigacp replied:
Ok cool. Someone should fix that doc, it still implies it needs a sentinel value even in the latest version.

yuzawa-san commented:
Can one of the project maintainers please remark on the outstanding comments?

saudet commented:
@Craigacp @yuzawa-san On an unrelated note, I'm pretty sure JavaCPP handles all of that out of the box. If there is anything missing with regards to thread safety, I would like to know about it, and will fix anything that needs to be fixed.

/cc @karllessard

yuslepukhin commented:
I took a cursory look at it and will spend more time. While Closable and AutoClosable implementation is a good idea, I would strongly object refcounting them. Making an std::shared_ptr in Java is not a good way to go. What is protecting from leaked references?
Smaller functions with clear resource ownership either on the object level or scoped to a function is a way to go.
Segmentation faults is a good thing. They let you know there is a bug and prevent from producing garbage results.
The whole product is written in C++ and we are not somehow afraid of segmentation faults. It should be possible to write correct code in modern Java w/o reference counting objects and masking errors.

---
In reply to: [437608734](https://github.com/microsoft/onnxruntime/pull/4329#issuecomment-437608734) [](ancestors = 437608734)

yuzawa-san commented:
@yuslepukhin I respectfully disagree with the `Segmentation faults is a good thing.` assessment. It is an absolute dealbreaker in Java development. I understand why segfaults exist, but it should not be an indication of exceptional behavior or improper use in a publicly available Java library. References to native objects do not leak out of the Java code per the safeguards in NativeObject.java.

Honestly, I do not know enough C++ to improve the JNI portion to use that shared_ptr thing. I'll let @Craigacp (the original author) discuss the original design and whether that is feasible. Also the current implementation appears to be in C rather than C++ so I don't know what level of effort to upgrade it would be.

My implementation is greatly inspired by the tensorflow implementation. My rationale was that if they use it, then it would probably be good for here.

saudet commented:
> I took a cursory look at it and will spend more time. While Closable and AutoClosable implementation is a good idea, I would strongly object refcounting them. Making an std::shared_ptr in Java is not a good way to go. What is protecting from leaked references?

We can easily keep a tally of live references to make sure nothing leaks. For example, @digiovinazzo contributed a small MXBean recently to do that easily with JavaCPP: https://github.com/bytedeco/javacpp/pull/413

> Smaller functions with clear resource ownership either on the object level or scoped to a function is a way to go.
> Segmentation faults is a good thing. They let you know there is a bug and prevent from producing garbage results.
> The whole product is written in C++ and we are not somehow afraid of segmentation faults. It should be possible to write correct code in modern Java w/o reference counting objects and masking errors.

Unfortunately, no, it does not seem possible. OpenJDK is mulling changes to the Java language exactly this purpose:
http://cr.openjdk.java.net/~rpressler/loom/loom/sol1_part2.html#scope-variables

> @yuslepukhin I respectfully disagree with the `Segmentation faults is a good thing.` assessment. It is an absolute dealbreaker in Java development. I understand why segfaults exist, but it should not be an indication of exceptional behavior or improper use in a publicly available Java library. References to native objects do not leak out of the Java code per the safeguards in NativeObject.java.

If by safeguards you mean the garbage collector, there is no guarantees that it will ever even run. A better analogy for GC would be as a last ditch effort.

> My implementation is greatly inspired by the tensorflow implementation. My rationale was that if they use it, then it would probably be good for here.

That code from TensorFlow is no longer being maintained. They've switched to JavaCPP exactly because no one wants to reimplement all that stuff, and/or have multiple versions of essentially the same thing that can't interoperate across libraries.

Craigacp commented:
I think @yuslepukhin is suggesting that reimplementing something that looks like a shared_ptr in Java is a bad idea, not that we should use it. I wrote the original JNI code in C as it's easier to understand and C++ induces a lot of complexity for what's a relatively simple interface (plus the C++ API seems to lag the C API).

However I agree strongly with @yuzawa-san, no Java program should ever be able to crash the JVM with a seg fault, and if the ORT Java API allows that currently then it's a bug and we should fix it. The Onnx Runtime is not guaranteed to be the only program running in that JVM, having a way to take the JVM down is extremely problematic.

There are two main parts to this PR (in addition to some smaller cleanups), the change which makes `OrtSession.close()` blocking or terminating on running inference jobs, and the reference counting of the native objects such as `OnnxTensor`, `OnnxMap` and `OnnxSequence`. The former is necessary for thread safety, and the latter is one alternative (we could synchronize around all the accessors and the close method which would provide similar guarantees about the native references). I don't think this is really true reference counting as the reference count is only ever held while a thread is in an accessor method, and the `Onnx*` objects are not cloneable or serializable, and so it's sole point is to prevent one thread calling close while another thread is in the `getValue` method. The current approach is slightly more concurrent than synchronizing on the object would be, though it's not clear to me how many threads are likely to access the same native object at once, and so if that's a real issue.

@saudet this project is smaller than TF, and I would like to be able to easily migrate it to use newer Java features such as Panama when they become available, so I don't think JavaCPP is an appropriate solution. Using JavaCPP here requires having JavaCPP taking over an already complex build system, adding additional runtime dependencies, changing the way the binaries are packaged, and then changing all of the code to use it's idioms. You've already forked this project to make a JavaCPP version anyway.

yuzawa-san commented:
I examined the differences between the old  https://github.com/tensorflow/tensorflow/blob/master/tensorflow/java/src/main/java/org/tensorflow/Session.java and the new https://github.com/tensorflow/java/blob/master/tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Session.java

I see the JavaCPP pointers were upgraded. How the pointers themselves are managed (longs vs JavaCPP) with JNI is not really what this PR is about. I added some small safety boilerplate code and just moved the pointers (the `long nativeHandle`) from the the subclasses to a common superclass. The question of JavaCPP usage is beyond the scope of this PR in my opinion and if @saudet or somebody else wants to take the time to upgrade it in the future that's fine by me. My main goal here is to prevent segfaults and JNI bugs in the short term.

The "wait until all active usages are done" mechanism (along with requireHandle and PointerScope to ensure liveness and safe cleanup) is still in the newer TF code. This is the main aspect of the PR. The thread safety is different as @Craigacp pointed out (Mine has less locking, more CAS). What I call a `referenceCount`, is called `numActiveRuns` in the TF code. I can rename if you feel strongly. If we feel strongly about more synchronizing, I'm open to discuss that.

saudet commented:
> @saudet this project is smaller than TF, and I would like to be able to easily migrate it to use newer Java features such as Panama when they become available, so I don't think JavaCPP is an appropriate solution. Using JavaCPP here requires having JavaCPP taking over an already complex build system, adding additional runtime dependencies, changing the way the binaries are packaged, and then changing all of the code to use it's idioms. You've already forked this project to make a JavaCPP version anyway.

If the idea with the official Java API of ONNX Runtime is merely to provide a thin layer on top of the C API, then yes it doesn't really matter where the JavaCPP wrappers end up. For example, OpenCV's official Java API still doesn't even have anything to package their libraries in JAR files, and thus nothing for Maven, and they don't care, but whatever, they are not providing anything that isn't already part of the C++ API anyway, so if it's the same here, sure. We can just keep bundling both Java and C/C++ APIs, and users that wish to get more features, friendlier interoperability, or higher speed can easily switch to the JavaCPP Presets at a later point in time.

BTW, please note that we haven't forked ONNX Runtime, we're merely providing more efficient and more usable packages.

/cc @EmergentOrder

Also, FYI, Panama has officially removed anything C++ from their roadmap, so I also do intend to use what they provide as part of JavaCPP--if it makes sense. Just using a shiny new API just because it's new and shiny doesn't make sense. It needs to provide something substantial, and right now, Panama is still a lot slower than JNI: https://github.com/zakgof/java-native-benchmark

> The "wait until all active usages are done" mechanism (along with requireHandle and PointerScope to ensure liveness and safe cleanup) is still in the newer TF code. This is the main aspect of the PR. The thread safety is different as @Craigacp pointed out (Mine has less locking, more CAS). What I call a `referenceCount`, is called `numActiveRuns` in the TF code. I can rename if you feel strongly. If we feel strongly about more synchronizing, I'm open to discuss that.

I see, ok, in that case then, it doesn't sound like you want to be using anything that resembles reference counting. Panama is having a lot of problems solving this exact kind of problem, and they've ruled out reference counting precisely because there is no way to make sure the counter always gets properly updated, without a whole lot of changes to the Java language that is:
https://mail.openjdk.java.net/pipermail/panama-dev/2020-May/008828.html
@yuslepukhin is right, reference counting isn't going to get you any kind of guarantee of correctness...

yuslepukhin commented:
I think @saudet has made all the points that I neglected to make clear so I do not need to add anything in principle.
Here is the main outtake. 

1. Re-counting as implemented has nothing to do with thread safety of both the native objects and its managed wrappers. So I suggest we make a change to the title of the PR. It has to do with managing objects life span which I am not convinced we need here. I can see we have the same attempt in a couple classes in our C# interfaces, I will correct it in due time.

2. AutoClosable interfaces are good since there is nothing else in Java that would provide timely release of native resources. There is a good question to ponder over: is there any connection between garbage collection and native objects? The correct answer, there is none. We can run out of the resources and native memory a thousand times and the garbage collector may still not run. Even if did, it would not release native resources bc it does not know anything about it.
Another equally good question, since we are on the topic, what is a good use of finalizers? The correct answer is there NONE. The should not have existed in the first place. Occasionally, in a DEBUG setting you can detect that a AutoClosable object was not timely closed, but it is not guarantee. Knowing this type of use is more beneficial in interview setting rather than in practice.
Finalizers do not run at the time of garbage collection, they run, if ever, at a different point in time and by  then we are already out of native memory, out of native resource handles and we are deadlocked if native synchronization objects are involved. Java lacks destructors and the closest equivalent for them is try/finally AKA AutoClosable objects now (better late than never).

Whether one disagrees with the value of Segmentation Fault or not in a particular setting, it is an unchangeable feature of the operating system. Any change we do should be aimed at the correctness, not running away from it. We certainly can not afford writing code that consumes CPU cycles merely to account for different levels of coding maturity.

yuzawa-san commented:
I've changed the nomenclature to stay away from reference counting. Here is a gist with a very barebones multithreaded environment: https://gist.github.com/yuzawa-san/e12b16a4aac7dc7bec2a94c4b8c2da89 This patch fixes the crash, so I'll update the title to reflect this.

Craigacp commented:
@yuslepukhin I agree that this is more about managing object lifetimes than thread safety, but it turns out the major issue for managing the object lifetimes is sharing an object between threads (in this case the OrtSession) and having one session accidentally close it out from under you. Everything else in the API pretty much just works when used by multiple threads, provided those threads don't call close. try-with-resources isn't the best solution in multi-threaded code like this, especially when the object is long lived (e.g. OrtSession), but it's all we've got which is why everything is already AutoCloseable. This PR adds another layer which alters the behaviour when closing OrtSession if another thread has a run call in flight, which prevents a JVM crash. Preventing JVM crashes is definitely worthwhile, as unlike C# the JVM doesn't really have any operations that can cause a segfault other than JNI so Java developers are not accustomed to seeing them.

saudet commented:
What about doing like @mcimadamore suggests on Panama's mailing list? I'm guessing something like this:
```java
OrtSession s = ...
s.withHandle(SessionHandle sh -> {
    sh.run(input)
});
s.close() // will fail (or wait) if there are any pending handles
```
The point is the OrtSession cannot be used without a handle, so all processing has to be done within the closure.
I'm not sure I agree with the usefulness of this, but this is coming from a Java compiler architect :)

yuzawa-san commented:
Some observations:
* Try-with-resources and closures seem pretty similar. I guess with the former we cannot guarantee the caller will close the Closeable versus with the closure we can put whatever code in the implementation of `withHandle` to bookend the call to the body of the closure. However, I'm not super concerned about that since I don't even want the everyday developer to know about checking out the handle. The NativeObject.use() which returns NativeUsage is package private so developers would not check out a reference. That is, OrtSession.run() internally calls NativeObject.use() and uses a try-with-resources in there. One value that closures has is that it contractually forces a certain usage rather than the little "internal developers should use NativeUsage in a try-with-resources" javadoc I currently have which puts it on the mercy of future code reviewers.
* I was playing around and the closure case gets weird in cases when you want to return from inside the closure.  I suppose we could make something like that impossible and you could pass in value to be populated in the closure:
```
final MyResult result = new ...
NativeObject s = ...
s.withHandle(Handle h -> {
    byte[] rawResult= h.run(input);
    result.setValue(rawResult);
});
// versus
try(Handle h = use()){
    return new MyResult(h.run(input));
}
```
I guess it is just a different way of thinking about it, but I don't know if it is worth the complexity. Also what if you want to make MyResult immutable?
* Also the functional interface gets a little odd when you throw different type of exceptions from within. It seems dirty to add throws Exception to the functional interface since it is overly broad and could be irrelevant in cases when nonesuch exceptions are present which forces the caller to handle them. Like what if h.run() throws OrtException and maybe new MyResult throws an IOException. Or what if the body of the never throws any Exceptions the caller would have to handle. Or perhaps the implementation of `withHandle` could wrap it in a subclass of RuntimeException, but then information is lost.
* If you did make the functional interface allowed return a value (or throw different exceptions) I guess you would have to consider the nested case where multiple handles are needed (like currently in OrtSession.run()) the return values would have to be propagated out and the throws must be consistent.
* I didn't want to break external API consistency for this library, but I guess given the infancy of this project maybe it could be an option.

I agree with you: definitely interesting, but I'm not fully sold. Honestly, I have not examined Panama in close detail, but I think the closure idea could work in certain cases. If you or somebody else want to do closures here in an elegant and proper way in a future PR, but I think I'll hold off for now.

saudet commented:
@yuzawa-san Don't misunderstand, I'm not recommending to force that kind of API on users, but it does work, for some things, just like try-with-resources works, for some things, and garbage collection works, for some things, and reference counting works, for some things (but there is no magic bullet that works for everything, and Panama isn't really interested in offering one either, so don't set your hopes on that front either). There is no way however to make reference counting thread-safe, unless we follow some magic enchantment protocol, so it doesn't make sense to force users into an API like that and tell them that it's thread-safe! It doesn't offer enough to be worth forcing users to use these things, and that's the point. With the JavaCPP wrapped C/C++ API, it's very easy to implement this kind of resource management on top, something like this:
```java
class SessionHolder {
    Session session = ...
    List<Session> list = ...
    Lock lock = ...
    public void withSession(SessionClosure closure) {
        lock.lock();
        try {
            if (session == null) // throw some exception
            Session s = new Session(session); // only copies the pointer address
            list.add(s);
            lock.unlock();
            closure.run(s);
            lock.lock()
            list.remove(s);
            s.setNull();
        } finally {
            lock.unlock();
        }
    }
    public void close() {
        lock.lock()
        try {
            if (!list.isEmpty()) {
                // wait somehow or throw exception
            }
            session.deallocate();
            session = null;
        } finally {
            lock.unlock();
        }
    }
}

SessionHolder h = new SessionHolder(...);

holder.withSession(Session s -> {
    s.run(input);
});
holder.close();
```

And we don't need to add that to ONNX Runtime or anything. Users can already do that or anything else on their own just fine. JavaCPP doesn't force anyone into doing things in one specific way. 

@Craigacp This is the kind of thing I'm talking about. You're offering users an API that isn't actually usable, and forcing them to open pull requests like this that are getting rejected. Why not offer them something usable, that they can build upon?

Craigacp commented:
@saudet The API is perfectly usable (because we and others are using it in production), this patch fixes an issue that I hadn't hit because I use the session in a try-with-resources and don't attempt to close it while run calls are in flight, and it doesn't require adding large numbers of locks or inducing new dependencies. I don't think exposing the closure version to users is particularly helpful, it's too low level. The purpose of reviews is to discuss things, and hopefully we're still making process on that front here. I'm not sure that you presupposing the outcome of the review when you haven't contributed anything to the project is particularly helpful.

I think the reference counting approach is sufficiently hidden that it will work in the OrtSession and prevent the crashes. It's not really counting references, it's counting the number of methods in flight using the native resource, and it's completely hidden from the user of the Java API. For the similar issues in OnnxValue subclasses then I think making the relevant methods synchronized is sufficient as they are immutable and it's not necessary to have concurrent access to their `getValue()` methods.

mcimadamore commented:
> It needs to provide something substantial, and right now, Panama is still a lot slower than JNI: https://github.com/zakgof/java-native-benchmark

Please check your facts. The benchmark you refer to is an year old, and is using a completely different Panama architecture (and API) compared to the one we are using now, for both memory and foreign function access. Out of curiosity I put together a small (unscientific) benchmark which does something similar to the one you linked (but using Linux, I used `localtime_r`) - Panama is just ~10% slower than pure JNI (I did not use JavaCPP); the only overhead I can see is associated to memory allocation/deallocation (compared to Unsafe::allocateMemory/freeMemory), where some extra logic is added so that memory access can be considered "safe" (as in avoiding hard JVM crashes due to segfaults). While we need to keep improving, these numbers are very different from those in the benchmark you linked (which shows Panama to be 20x slower than JNI/JavaCPP).

yuzawa-san commented:
We could debate the nuances of the internal implementation (which was done prior to my arrival) ad infinitum and yes there are philosophical differences between segfaults in C/CPP-world and the somewhat-equivalent NullPointerException in Java that could be re-litigated, but I would like to get this underlying bug fixed, as it is a blocker to my (and likely others') future use of this API. The main objective here was to fix the defect shown here which @Craigacp mentioned they had overlooked in the original implementation: https://gist.github.com/yuzawa-san/e12b16a4aac7dc7bec2a94c4b8c2da89 I did not alter the Java public facing API with the exception of adding the OrtSession.terminate() method, which I documented. What are the next steps to getting this approved?

hariharans29 commented:
/azp run Linux CPU CI Pipeline,Linux CPU x64 NoContribops CI Pipeline,Linux GPU CI Pipeline,Linux GPU TensorRT CI Pipeline,MacOS CI Pipeline,MacOS NoContribops CI Pipeline,Windows CPU CI Pipeline,Windows GPU CI Pipeline,Windows GPU TensorRT CI Pipeline

hariharans29 commented:
/azp run orttraining-linux-ci-pipeline,orttraining-mac-ci-pipeline,orttraining-linux-gpu-ci-pipeline,centos7_cpu,Linux OpenVINO CI Pipeline

hariharans29 commented:
@Craigacp  @yuslepukhin - is this change good to go ?

saudet commented:
> @saudet The API is perfectly usable (because we and others are using it in production), this patch fixes an issue that I hadn't hit because I use the session in a try-with-resources and don't attempt to close it while run calls are in flight, and it doesn't require adding large numbers of locks or inducing new dependencies. I don't think exposing the closure version to users is particularly helpful, it's too low level. The purpose of reviews is to discuss things, and hopefully we're still making process on that front here. I'm not sure that you presupposing the outcome of the review when you haven't contributed anything to the project is particularly helpful.

I have been trying to contribute, but with your authority here as representative from Oracle you are preventing any potential contributions that I have to offer. We need something like JavaCPP for performance reasons and to interoperate with other libraries like OpenCV and TensorFlow. (BTW, you do know that OpenCV is probably the second most used AI-related framework after TensorFlow, right? Why isn't Oracle doing anything about that one?) The only reasons I have ever heard from you about why you are refusing to consider using (or even creating yourself something similar to JavaCPP), is what you said above, but you provide no specifics:

> Using JavaCPP here requires having JavaCPP taking over an already complex build system, adding additional runtime dependencies, changing the way the binaries are packaged, and then changing all of the code to use it's idioms

Where exactly is the build "taken over"? What is "overly complex" about a ~100 lines Bash script and Java config file? What additional dependencies? How do you want to package the binaries to have them share stuff with OpenCV and TensorFlow? The way this sounds to me is you're saying that everything the Python community does with tools like conda, pip, setuptools, etc is wrong, and Java developers shouldn't use tools like that. If that's not what you mean, then at least please offer an alternative! Remember that you're speaking for Oracle here.

> I think the reference counting approach is sufficiently hidden that it will work in the OrtSession and prevent the crashes. It's not really counting references, it's counting the number of methods in flight using the native resource, and it's completely hidden from the user of the Java API. For the similar issues in OnnxValue subclasses then I think making the relevant methods synchronized is sufficient as they are immutable and it's not necessary to have concurrent access to their `getValue()` methods.

That's exactly the problem, it's all hidden, it prevents users from doing things in whatever way they may want to do things!

saudet commented:
> @saudet please check your facts. The benchmark you refer to is an year old, and is using a completely different Panama architecture (and API) compared to the one we are using now, for both memory and foreign function access. Out of curiosity I put together a small (unscientific) benchmark which does something similar to the one you linked (but using Linux, I used `localtime_r`) - Panama is just ~10% slower than pure JNI (I did not use JavaCPP); the only overhead I can see is associated to memory allocation/deallocation (compared to Unsafe::allocateMemory/freeMemory), where some extra logic is added so that memory access can be considered "safe" (as in avoiding hard JVM crashes due to segfaults). While we need to keep improving, these numbers are very different from those in the benchmark you linked (which shows Panama to be 20x slower than JNI/JavaCPP).

That's great news! Thanks for the update. I'm looking forward to the day when Panama actually stops using JNI though.

saudet commented:
> We could debate the nuances of the internal implementation (which was done prior to my arrival) ad infinitum and yes there are philosophical differences between segfaults in C/CPP-world and the somewhat-equivalent NullPointerException in Java that could be re-litigated, but I would like to get this underlying bug fixed, as it is a blocker to my (and likely others') future use of this API. The main objective here was to fix the defect shown here which @Craigacp mentioned they had overlooked in the original implementation: https://gist.github.com/yuzawa-san/e12b16a4aac7dc7bec2a94c4b8c2da89 I did not alter the Java public facing API with the exception of adding the OrtSession.terminate() method, which I documented. What are the next steps to getting this approved?

It's not just a bug fix (a bug fix would just throw an exception or something from somewhere), it forces users to adopt a way of doing things that purports to be thread-safe, but that really isn't, and that will slow down processing. If there were a way to disable all of this, I would be fine with these changes though. Is there a way to disable all of this? (I also think that kind of feature should be optional in TensorFlow. I'll probably get to it when I need to hack that there. The developer that used to work at Google on this API wasn't an expert in Java and I don't think his code got reviewed.)

Craigacp commented:
> I have been trying to contribute, but with your authority here as representative from Oracle you are preventing any potential contributions that I have to offer. We need something like JavaCPP for performance reasons and to interoperate with other libraries like OpenCV and TensorFlow. (BTW, you do know that OpenCV is probably the second most used AI-related framework after TensorFlow, right? Why isn't Oracle doing anything about that one?) The only reasons I have ever heard from you about why you are refusing to consider using (or even creating yourself something similar to JavaCPP), is what you said above, but you provide no specifics:

@saudet As I've told you several times before, I work in Oracle Labs, and do not speak for the company or the Java org in any way. In this particular case I'm speaking as the person who wrote the Java API for this project. I've not written anything for OpenCV for two reasons: one there is already a Java API, and two I don't work on computer vision. I'm unclear why the ORT library would need to interop with Tensorflow. If you want to use a TF model then export it in onnx format and compose it using the mechanisms already available. If there are performance problems that need addressing then open an issue with some numbers and I'll take a look.

> > Using JavaCPP here requires having JavaCPP taking over an already complex build system, adding additional runtime dependencies, changing the way the binaries are packaged, and then changing all of the code to use it's idioms
> 
> Where exactly is the build "taken over"? What is "overly complex" about a ~100 lines Bash script and Java config file? What additional dependencies? How do you want to package the binaries to have them share stuff with OpenCV and TensorFlow? The way this sounds to me is you're saying that everything the Python community does with tools like conda, pip, setuptools, etc is wrong, and Java developers shouldn't use tools like that. If that's not what you mean, then at least please offer an alternative! Remember that you're speaking for Oracle here.

Using JavaCPP requires having JavaCPP build the native layer, opting into JavaCPP's mechanisms for binary distribution, and you need to get the right set of dependencies. As we've found out with the recent issues in the TF Java project, the way that JavaCPP works can be confusing for some developers and it can be tricky to figure out the dependencies. I don't think it's worthwhile to switch to something more complex when we have something that works and is simpler for users of the library.

None of your points are relevant to this discussion, and adopting JavaCPP wouldn't fix the race condition between object use and closing here, given it's essentially a problem with the C API for the library (as calling the C API's close method while there is a run call in flight shouldn't segfault the process).

yuslepukhin commented:
I took a look at the PR and the scenario. While the scenario is valid, I do not feel that this specific user scenario complexity should be pushed down to the API level. A simple inference would have to pay for all refocunting and synchronization mechanisms.
In the scenario described, it is very important to get multi-threaded details correct. yet, the code silently allows destroying objects twice and other things which would not lead to good results, basically swallowing programming mistakes.

yuslepukhin commented on java/src/main/java/ai/onnxruntime/NativeObject.java:
```diff
@@ -0,0 +1,211 @@
+    synchronized (handleLock) {
+      if (closed) {
+        return;
+      }
```
This should throw some exception that would clearly tell that things are going wrong. Somehow, a thread is operating on an object that was already be destroyed and but it thinks it has a valid handle on it. With property implemented refcounting it should never happen, as it looks like the a thread is holding a reference to it but the object is destroyed anyway. This condition should prompt an assert or an non-recoverable exception. Which is what Segfault is currently doing for you.

Refcounting or not, each native object should have a proper and clear ownership and judjing but this piece of code, you do not expect that refcounting would actually save you. Why introducing it?
You can wrap a session object into your recounted class and save other users from the bearing the expense of it.

yuzawa-san replied:
The close() method is called by developers. It is within reason that somebody may close the object more that once and we do not really care since the object was closed already. An exception is thrown if the session is used again (like if they use run() again).

The active usage (NOT reference) counting is beside the point here, since that is done in private implementation hidden from the users and with try-with-resources.

yuslepukhin replied:
close() is usually the last thing that is being called on the object. Since the calling code seems to be unware that the underlying native object is no longer there, it can call all sorts of functions on it before close(). By this reasoning, one would need to check of the object is still around in every interface which is not being done. So it would still lead you to a segfault, would it not? Discovering the the object is gone at the time of close() does nothing to save you if the program continues to be invalid.  If there is not a clear ownership and the recounting is not implemented right such checks actually do nothing. If two threads are trying to close the object but the recount is 1, both would pass the check, and they would happily destroy the object twice thereby corrupting memory. And you already know where it leads -SEGFAULT.
Let me summarize: this PR introduces atomics and locks, both of which impose additional usage requirements and performance costs on other users for the sake of your usage scenario. However, I feel, that the additional overhead and complexity should be shifted to your code.
On another note, there are inter dependencies on some objects. You will see in the next release, OrtAllocators API being exposed and IOBidning. Those depend on Session. Session depends on Environment. Refcounting them individually would not make sense.

---
In reply to: [471666137](https://github.com/microsoft/onnxruntime/pull/4329#discussion_r471666137) [](ancestors = 471666137)

Craigacp replied:
The semantics of Java's AutoCloseable interface say:
```Note that unlike the close method of Closeable, this close method is not required to be idempotent. In other words, calling this close method more than once may have some visible side effect, unlike Closeable.close which is required to have no effect if called more than once. However, implementers of this interface are strongly encouraged to make their close methods idempotent.```
So having it accept multiple calls to close is idiomatic.

Craigacp replied:
However the upcoming changes to allocators and IObinding do make me worried we'll need a bigger solution to this. The allocators were originally exposed and had a more complicated tracking mechanism, but during the initial integration of the Java API we decided that it wasn't necessary and lots of complexity was removed by making the allocators an internal detail.

yuslepukhin replied:
Thanks for the correction. The underlying Release*() methods  are not idempotent. Nor, I feel, they are correctly made such by this PR. Perhaps, Closable is a better interface to implement.

---
In reply to: [471682951](https://github.com/microsoft/onnxruntime/pull/4329#discussion_r471682951) [](ancestors = 471682951)

yuslepukhin replied:
Yes, the allocators we are exposing actually hang of the session. They are dependent on the build and the execution providers present so one can allocate a native piece of memory and bind it to input/output. 
So, facing this user scenario, I would probably refcount the session object so I can swap it and make threads that are just starting they work work with the new session while others are still finishing up with the old one. But everything else, actually depends on the active usage and they way code is organized, contrary to the assumption made in this PR.
In C# we use a trick to add everrything IDisposable to a list so a single object is disposed with everything else inside. So each thread, while sharing a session, can create/destroy other native objects within a single run.
Again, this is a usage scenario. Keeping the API simple allows to face future with confidence.

---
In reply to: [471687336](https://github.com/microsoft/onnxruntime/pull/4329#discussion_r471687336) [](ancestors = 471687336)

Craigacp replied:
In Java `AutoCloseable` recommends idempotence, `Closeable` requires it. `Closeable` is explicitly documented that calling it twice has no effect.

yuzawa-san replied:
The increment/decrement and synchronization mechanisms would start to show degraded performance in cases where there is great thread contention, which is pretty expected. There is no free lunch when it comes to concurrency as always. With regards to the amount of CPU time this introduces for the simple single-threaded vanilla implementation like in the sample code, since that is a single thread, there should be very very minimal performance hit. The AtomicInteger updates for the active usage counts should be super quick since those are two CAS operations. Since there is only one thread, the CAS is mathematically guaranteed to be successful on the first attempt since no other thread would have attempted to update the count. The synchronized blocks are only relevant on close in the thread doing the close() and the last currently active run which notifies the waiting close thread. It is not a common use case to call close() from many threads at once.

I think that such small investments in processing are worthwhile to ensure safety. To reiterate, the TF project does a similar thing, so if they do it, I think there is enough of a community use case to warrant adding such safety mechanisms.

Craigacp replied:
Is there a description of the allocator semantics in the coming updates? I couldn't find a good one when I initially wrote the API, it seemed to be an undocumented facet of the C++ implementation and my C++ isn't good enough to figure out the behaviour from that. Plus undocumented implementation details are very different from documented specs.

Ditto for IOBinding.

yuslepukhin replied:
The documentation gap is recognized internally and there will be an effort to close it. Please, do ask me questions about the behavior.
Please, refer to this PR for more details. 
It also demonstrates the way the API is exposed in C#. [https://github.com/microsoft/onnxruntime/pull/4646](https://github.com/microsoft/onnxruntime/pull/4646)

---
In reply to: [471708975](https://github.com/microsoft/onnxruntime/pull/4329#discussion_r471708975) [](ancestors = 471708975)

yuslepukhin replied:
The initial allocator exposed was a default CPU allocator. It was a singleton that never required destruction.
The allocators obtained with the new API CreateAllocator() are wrapping instances that require ReleaeAllocator() call when no longer needed. 
Those refer to the internal Session allocators. 
The internal sessions allocators are tied to the execution providers.  Those depend on two factors: 1) whether it is present in the binary 2) whether it was added in the session options.
For example, if you have a build with CPU and CUDA, you specify then in the SessionOptions, create a Session.
Then you create an instance of OrtMemoryInfo that has all the characteristics of the allocator required. (Note, this is also a native Closable/Disposable class)
Using that you call CreateAllocator() which internally checks if the requested allocator is present within the session. It would then create a wrapper and return it.
You now can allocate memory on the device which returns OrtMemoryAllocation instance.
 This is useful in some scenarios 1) You do not want to reallocate memory for input/output and just re-use a  pre-allocated chunk progressively specifying bigger shapes but re-use the same chunk
2) You want to feed output back to input. So you just re-bind input/output trading places.
3) You have a dynamically shaped output. In this case you can not pre-allocated, but by using OrtMemoryInfo you can specify where to allocate and it will do so.
This are especially useful, if you need it to be on the device or you want to feed input that is already allocated, say on CUDA. So no copying to host memory is needed.
in cases 1 and 2 the caller owns the memory, in case of 3, Onnxruntime owns that (Returned OrtValue does).

---
In reply to: [471751379](https://github.com/microsoft/onnxruntime/pull/4329#discussion_r471751379) [](ancestors = 471751379,471708975)

saudet replied:
> 3) You have a dynamically shaped output. In this case you can not pre-allocate, but by using OrtMemoryInfo you can specify where to allocate and it will do so.
> This is especially useful, if you need it to be on the device or you want to feed input that is already allocated, say on CUDA. So no copying to host memory is needed.
> in cases 1 and 2 the caller owns the memory, in case of 3, Onnxruntime owns that (Returned OrtValue does).

@Craigacp Here you go, you should consider your Java API not supporting CUDA memory as a "performance problem". No need to run benchmarks to know that copying data around unnecessarily is going to be slow. And, no, the official Java API of OpenCV doesn't support CUDA, so even if you figure out something for ORT, it's still going to be a problem! Why do you refuse the recognize the importance of this? I'm not saying that JavaCPP is perfect, but at least I'm trying to do something about it, while you are not.

yuslepukhin replied:
1) Yes, the allocators are thread-safe. They are the same ones used by the session object and the session object support concurrent execution my multiple threads.
2) Yes, it is possible to bind inputs/outputs to a slub of memory allocated by someone else. When we bind a pre-allocated piece of memory, it is not owned by ORT or the corresponding OrtValue. One of the python binding tests makes use of CUDA memory allocated by PyTorch. Current C# API requires passing OrtMemoryAllocation wrapping the chunk. That currently require an allocator ptr. So I can see that in C# we could make it more convinient to supply 3rd party allocations.

---
In reply to: [474764836](https://github.com/microsoft/onnxruntime/pull/4329#discussion_r474764836) [](ancestors = 474764836)

yuslepukhin commented:
Nope. We can not accept this change.

---
In reply to: [673709909](https://github.com/microsoft/onnxruntime/pull/4329#issuecomment-673709909) [](ancestors = 673709909)

yuzawa-san commented:
I'm going to port the non-controversial code cleanups into a new PR.

saudet commented:
> Using JavaCPP requires having JavaCPP build the native layer, opting into JavaCPP's mechanisms for binary distribution, and you need to get the right set of dependencies. As we've found out with the recent issues in the TF Java project, the way that JavaCPP works can be confusing for some developers and it can be tricky to figure out the dependencies. I don't think it's worthwhile to switch to something more complex when we have something that works and is simpler for users of the library.

@Craigacp FYI, we can easily integrate JavaCPP into other build systems, using it only to generate code. I've just done it successfully for Android Studio, and that uses Gradle to call CMake:
 * https://github.com/bytedeco/gradle-javacpp#integration-with-android-studio

Without any explanations about why writing JNI code manually is less "complex" than using a code generator that does all that for us, your statements are simply not true!
